test/cases/config-parser.c

```diff
@@ -0,0 +1,617 @@
+#include "test.h"
+#include "test-tempfile.h"
+#include "test-assertions.h"
+
+#include "charset.h"
+
+#include "config-parser.h"
+
+static const char * const test_config_file_content =
+	"[ducks]\n"
+	"colour = brown\n"
+	"count = 7\n"
+	"weight = 64 lb.\n";
+
+#define ARRANGE_TEST_CONFIG_FILE(CFG) \
+	do \
+	{ \
+		char test_config_file[TEST_TEMP_FILE_NAME_LENGTH]; \
+\
+		if (!test_temp_file(test_config_file, test_config_file_content, -1)) \
+			return SCHISM_TESTRESULT_INCONCLUSIVE; \
+\
+		int init_result = cfg_init(&cfg, test_config_file); \
+\
+		if (init_result < 0) \
+		{ \
+			test_log_printf("cfg_init(\"%s\") failed with result %d\n", test_config_file, init_result); \
+			RETURN_FAIL; \
+		} \
+	} while (0)
+
+testresult_t test_config_file_defined_values(void)
+{
+	// Arrange
+	cfg_file_t cfg;
+
+	ARRANGE_TEST_CONFIG_FILE(cfg);
+
+	// Act
+	const char *colour = cfg_get_string(&cfg, "ducks", "colour", NULL, 0, "abcd");
+	int count = cfg_get_number(&cfg, "ducks", "count", 1234);
+	int weight = cfg_get_number(&cfg, "ducks", "weight", 1234);
+
+	// Assert
+	ASSERT(strcmp(colour, "brown") == 0);
+	ASSERT(count == 7);
+	ASSERT(weight == 64);
+
+	cfg_free(&cfg);
+
+	RETURN_PASS;
+}
+
+testresult_t test_config_file_undefined_values_in_defined_section(void)
+{
+	// Arrange
+	cfg_file_t cfg;
+
+	ARRANGE_TEST_CONFIG_FILE(cfg);
+
+	// Act
+	const char *sauce = cfg_get_string(&cfg, "ducks", "sauce", NULL, 0, "soy");
+	int feathers = cfg_get_number(&cfg, "ducks", "feathers", 94995);
+
+	// Assert
+	ASSERT(strcmp(sauce, "soy") == 0);
+	ASSERT(feathers == 94995);
+
+	cfg_free(&cfg);
+
+	RETURN_PASS;
+}
+
+testresult_t test_config_file_undefined_section(void)
+{
+	// Arrange
+	cfg_file_t cfg;
+
+	ARRANGE_TEST_CONFIG_FILE(cfg);
+
+	// Act
+	const char *weather = cfg_get_string(&cfg, "barbecue", "weather", NULL, 0, "elf");
+	int dismal = cfg_get_number(&cfg, "barbecue", "dismal", 758);
+
+	// Assert
+	ASSERT(strcmp(weather, "elf") == 0);
+	ASSERT(dismal == 758);
+
+	cfg_free(&cfg);
+
+	RETURN_PASS;
+}
+
+testresult_t test_config_file_obviously_broken_values(void)
+{
+	// Arrange
+	cfg_file_t cfg;
+
+	ARRANGE_TEST_CONFIG_FILE(cfg);
+
+	// Act
+	// could be separate tests...
+	const char *str_null_section = cfg_get_string(&cfg, NULL, "shouldn't crash", NULL, 0, "ok");
+	const char *str_null_key = cfg_get_string(&cfg, "shouldn't crash", NULL, NULL, 0, "ok");
+	int num_null_section = cfg_get_number(&cfg, NULL, "shouldn't crash", 1);
+	int num_null_key = cfg_get_number(&cfg, "shouldn't crash", NULL, 1);
+	const char *str_null_default_value = cfg_get_string(&cfg, "doesn't", "exist", NULL, 0, NULL);
+
```

Comment:
as per b11e3dd9ab191f20830062be74f6035af4db78ba we should have a test that supplies a valid buffer but puts in an (obviously bad) 0 for the length